### DISCUSSION TITLE
Obtain custom params in RTMP urls using the API

### DISCUSSION THREAD

**Discussion by danrossi** [OvenMediaLabs/OvenMediaEngine#1648] [Help] [answered]
I have a group call backend that can support signalling RTMP streams to be added to the room. But to do so I need to encode data like title and roomname as the stream name. This makes it hard for policy token generation having to us an encoded stream name. If there is a way to obtain params added to the rtmp urls it might help. 

I have to check active streams with 

/v1/vhosts/default/apps/app/streams

But this doesn't give me detail about an rtmp stream hence having to encode in the name,

**Comment by bchah** [comment]
You could try using AdmissionWebhooks which provides the full incoming URL to your own server: https://airensoft.gitbook.io/ovenmediaengine/access-control/admission-webhooks

**Comment by danrossi** [comment]
webhooks to the same service as the websocket server may work to notify publish has begun. Or is there another webhook for notifying publishing has begun ?  So there is no way to get rtmp params using the rest api of the available streams ?

**Reply by bchah** [reply, marked as answer]
The webhook can be sent anywhere you choose. You're correct that the API does not store the URL parameters, just the names of the active streams themselves. However you could use the webhook to store the parameters in your own DB or memory when the stream starts and reference that value later.

**Reply by danrossi** [reply]
I'm testing out the webhook to my "rooms" node server. I have to figure out how to replicate app instances like wowza that partition streams into instances on the app. I'm wondering if I can obtain params from the webrtc streams this way so I know which room its from.  So I can use a webhook as a publish notification also. Or else I have to include the room name in the stream name which complicates creating the tokens which requires the entire url and stream name.

**Reply by bchah** [reply]
AdmissionWebhooks let you provide a 'new_url'. So use a unique identifier as a sort of 'connection key' and store your room info against that in your own database, so there is no need to transmit it in the key itself. Then provide the real stream key as the new url and make whatever calls you need based on what's in the database.

**Reply by danrossi** [reply]
There is no example how to verify X-OME-Signature for the webhook ? I'll figure out something. The webhook is used as a publish notifier which will notify a room map. So the original connection needs to specify the room name to map to. I hope it can be obtain from params in webrtc like it can for RTMP.

**Reply by bchah** [reply]
It works similarly to SignedPolicy, check out the example code in misc for how that is generated (HMAC-SHA1 the url or payload, then base64 that).

Or you can just bypass use of the signature and only allow requests from the local host assuming your other endpoint is running on the same server.

**Reply by danrossi** [reply]
So I compare encrypted signature rather than use jwt api to verify ? What is the payload used to also create a signature on the control server side ? This is using the policyurl. 

https://github.com/AirenSoft/OvenMediaEngine/blob/master/misc/signed_policy_url_generator.js#L38

**Reply by bchah** [reply]
The payload creating X-OME-Signature for AdmissionWebhooks is the body of the HTTP request itself: https://airensoft.gitbook.io/ovenmediaengine/access-control/admission-webhooks#security

**Reply by danrossi** [reply]
This is a node example backend I am building but will do one for python also. This request is made before publish is made so further checks is needed if it's publishing to signal to subscribe to the stream. However the url with params is shown in the body which is useful. 

```
app.post('/v1/admission', (req, res) => {

  
    const token = req.headers['x-ome-signature'];

  
    if (token == null) return res.sendStatus(401)

    const body = req.body;


    let signature = base64url(crypto.createHmac('sha1', options.secret).update(JSON.stringify(body)).digest());

    console.log("Token: ", token);
    console.log("Body: ", body);
    console.log("Sig: ", signature);

    if (token !== signature) return res.sendStatus(401);


    res.send( JSON.stringify({ allowed: true }) );


  
  });
  
  ```

**Reply by danrossi** [reply]
I'm sorry is there a webhook notification for when an RTMP publish stops ? It will help notifications to stop subscribing to it.

**Reply by bchah** [reply]
There is - it would have a 'closing' status and an 'incoming' direction. However this only gets sent if the RTMP client does a proper disconnect. Sudden disconnects don't count (and are common e.g. crash, hardware power off). So you should periodically poll the OME API for a list of streams and check there to be sure.

FYI, I've never been able to get NodeJS to give me an accurate HMAC-SHA1 hash. Something to do with the way HMAC works there. Let me know if you have better luck.

**Reply by danrossi** [reply]
I have formed this request handler and noticed some issues. The webhook is pre publish not post publish so I have to also run an api check after to check if the stream is available.

I only want the admission webhook on the app used for rooms, I am going to publish a mixed stream back to a seperate application. So a per application config rather than global would be good. And it should send the app and streamname in the payload to prevent having to parse the urls ? 

```
function checkStreamActive(app, streamName, room) { 
  
  const path = `/v1/vhosts/default/apps/${app}/streams/${streamName}`;

  //base64 encode token

  const response = (res) => {
    console.log(`statusCode: ${res.statusCode}`)
    res.on('data', (data) => {

      let body = JSON.parse(data);

      switch (body.statusCode) {
        case 200:
          //notify participants stream is available
          room.participantStreamAvailable(streamName);
        break;
        default:
          console.log(`Stream ${streamName} Not Active Checking Again`);
          setTimeout(checkStreamActive, 1500, app, streamName, room);

        break;
      }
    })
  };

  apiCommand(path, response);
}
app.post('/v1/admission', (req, res) => {

  
    const token = req.headers['x-ome-signature'];

  
    if (token == null) return res.sendStatus(401)

    const body = req.body;


    let signature = base64url(crypto.createHmac('sha1', options.secret).update(JSON.stringify(body)).digest());

    console.log("Token: ", token);
    console.log("Body: ", body);
    console.log("Sig: ", signature);

    if (token !== signature) return res.sendStatus(401);

    const request = body.request,
    isPublishing = request.direction == "incoming";

    if (isPublishing) {

      const isClosing = request.status == "closing";

      const url = new URL(request.url);

      //console.log("URL", url);

      const paths = url.pathname.split("/"),
      app = paths[1],
      streamName = paths[2],
     

      if (app == roomsAop) {
        const roomName = url.searchParams["room"];

        switch (request.protocol) {
          case "WebRTC":
            console.log("WebRTC");
            checkStreamActive(app, streamName, room);  
          break;
        }
      }
    
      
    }

    


    res.send( JSON.stringify({ allowed: true }) );


  
  });
  ```

**Reply by bchah** [reply]
You can create multiple VirtualHosts and use AdmissionWebhooks on just one of them, or you can add your own logic to deal with different destinations based on the application in the URL. You can definitely create the desired outcome one of these ways.
